apps/opik-backend/src/main/java/com/comet/opik/domain/attachment/AttachmentStripperService.java

```diff
@@ -353,53 +364,144 @@ private void wrapWithSpanVoid(String spanName, Runnable action) {
     }
 
     /**
-     * Processes all base64 strings found in a JSON string and replaces them with attachment references.
+     * Recursively processes a JSON node, checking each text field for potential attachments.
      *
-     * Uses regex to find base64 strings longer than the minimum threshold, processes each one
-     * as a potential attachment, and replaces valid attachments with reference strings.
+     * This method walks the JSON tree without converting to string, which is much more efficient
+     * for large payloads. It handles:
+     * - Object nodes: recursively processes each field
+     * - Array nodes: recursively processes each element
+     * - Text nodes: checks if content matches base64 pattern and processes as attachment
+     * - Other nodes: returns as-is (numbers, booleans, null, etc.)
+     *
+     * @return the processed JsonNode (may be the same instance if no attachments found)
      */
+    private JsonNode processJsonNode(
+            JsonNode node,
+            AtomicInteger attachmentCounter,
+            AttachmentContext ctx) {
 
-    private String processBase64InJsonString(String jsonString,
-            UUID entityId,
-            EntityType entityType,
-            String workspaceId,
-            String userName,
-            String projectName,
-            String context) {
+        if (node == null || node.isNull()) {
+            return node;
+        }
+
+        return switch (node.getNodeType()) {
+            case OBJECT -> processObjectNode(node, attachmentCounter, ctx);
+            case ARRAY -> processArrayNode(node, attachmentCounter, ctx);
+            case STRING -> processTextNode(node, attachmentCounter, ctx);
+            default -> node; // Numbers, booleans, null, etc. pass through unchanged
+        };
+    }
 
-        // Early exit: if the entire JSON string is smaller than our minimum base64 size,
-        // it can't possibly contain any base64 attachments worth processing
-        if (jsonString.length() < minBase64Size) {
-            return jsonString;
+    /**
+     * Processes a JSON object node by recursively processing each field.
+     *
+     * @return the processed object node, or the original if no changes were made
+     */
+    private JsonNode processObjectNode(
```

Comment:
Can we add some instrumentation at the point we believe it could be more CPU-intensive?